File: oop-exercises/inheritance/nestjs/test/entity/CreditCardAccount.test.ts
import { CreditCardAccount } from "../../src/entity/CreditCardAccount";

describe('CreditCardAccount', () => {
  it('should create credit card account with zero balance', () => {
    const creditCardAccount = new CreditCardAccount(0, 0.5);

    expect(creditCardAccount.getBalance()).toEqual(0);
    expect(creditCardAccount.getInterestRate()).toEqual(0.5);
    expect(creditCardAccount.calculateInterest()).toEqual(0);
  });

  it('should create credit card account with some balance', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    expect(creditCardAccount.getBalance()).toEqual(100);
    expect(creditCardAccount.getInterestRate()).toEqual(0.5);
    expect(creditCardAccount.calculateInterest()).toEqual(50);
  });

  it('should add deposit to the account', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.deposit(100);

    expect(creditCardAccount.getBalance()).toEqual(200);
  });

  it('should withdraw', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.withdraw(50);

    expect(creditCardAccount.getBalance()).toEqual(50);
  });

  it('should throw an error if dont have sufficient balance', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    expect(() => {
      creditCardAccount.withdraw(200);
    }).toThrowError('Insufficient balance to perform the transaction.');
  });

  it('should charge', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.charge(50);

    expect(creditCardAccount.getBalance()).toEqual(50);
  });

  it('should throw an error if dont have sufficient balance when charging', () => {
    const creditCardAccount = new CreditCardAccount(100, 0.5);

    expect(() => {
      creditCardAccount.charge(200);
    }).toThrowError('Insufficient balance to perform the transaction.');
  });
});

File: oop-inheritance/spring-boot/src/test/java/com/juandaabril/entity/CreditCardAccountTest.java
package com.juandaabril.entity;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import org.junit.jupiter.api.Test;

class CreditCardAccountTest {

  @Test
  public void shouldCreateCreditCardAccountWithZeroBalance() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(0, 0.5);

    assertThat(creditCardAccount.getBalance()).isEqualTo(0);
    assertThat(creditCardAccount.getInterestRate()).isEqualTo(0.5);
    assertThat(creditCardAccount.calculateInterest()).isEqualTo(0);
  }

  @Test
  public void shouldCreateCreditCardAccountWithSomeBalance() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    assertThat(creditCardAccount.getBalance()).isEqualTo(100);
    assertThat(creditCardAccount.getInterestRate()).isEqualTo(0.5);
    assertThat(creditCardAccount.calculateInterest()).isEqualTo(50);
  }

  @Test
  public void shouldAddDepositToTheAccount() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.deposit(100);

    assertThat(creditCardAccount.getBalance()).isEqualTo(200);
  }

  @Test
  public void shouldWithDraw() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.withdraw(50);

    assertThat(creditCardAccount.getBalance()).isEqualTo(50);
  }

  @Test
  public void shouldThrowAnErrorIfDontHaveSufficientBalance() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    assertThatThrownBy(
            () -> {
              creditCardAccount.withdraw(200);
            })
        .hasMessage("Insufficient balance to perform the transaction.");
  }

  public void shouldCharge() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    creditCardAccount.charge(50);

    assertThat(creditCardAccount.getBalance()).isEqualTo(50);
  }

  @Test
  public void shouldThrowAnErrorIfDontHaveSufficientBalanceWhenCharging() {
    CreditCardAccount creditCardAccount = new CreditCardAccount(100, 0.5);

    assertThatThrownBy(
            () -> {
              creditCardAccount.charge(200);
            })
        .hasMessage("Insufficient balance to perform the transaction.");
  }
}
